### PR TITLE
Attempt to speed up integration tests using pre-built images

### PR DESCRIPTION
**Pull request by chrisgilmerproj**
## Description

This changes the e2e tests in CircleCI to use docker-compose which will use pre-built images from AWS ECR to speed up testing. My tests show that this takes us from about 16 minutes for e2e tests to as low as 10 minutes (depending on how CircleCI is queuing containers). 

Further speed ups could happen if we balance how many tests exist in each cypress test file. You can compare how big each test file is by using `find cypress/integration/ -type f  | xargs -I {} wc -l {} | sort`. The mothballing of HHGs is going to remove all the TSP and some Office tests but we could still improve this a bit.

Also in this PR:
- Removing docker-layer-caching in CircleCI to get a 30+ second speed up on steps that don't require it
- The local `Dockerfile.migrations_local` more closely matches the `Dockerfile.migrations` and anything extra was moved to `Dockerfile.tools` (meaning we have higher fidelity in testing)
- All the old Makefile targets for the e2e tests that touched the `test` database have been removed
- An error was caught were setting `MIGRATION_PATH` with multiple `;;` semicolons (ie empty path) would break the migrations.
- docker-compose now uses port 4000 similar to cypress tests. That's because cypress doesn't make it easy to modify the expected port for these tests.

## Setup

Running locally should be unaffected:

```sh
make e2e_test
make e2e_test_docker
```

Also, this change maintains the promise that what is run locally matches what will be run in CircleCI.

## Code Review Verification Steps

* [x] Request review from a member of a different team.

### CODE REVIEW

**Comment by chrisgilmerproj** [comment]
I've done a little looking here and honestly this would probably work better if I used docker-compose.

File: cypress/support/constants.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ export const longPageLoadTimeout = 10000;
 export const fileUploadTimeout = 10000;
 
 // Base URLs
-const cypressClientPort = 4000; //change this to 3000 to run cypress against dev instance
+const cypressClientPort = process.env.CYPRESS_CLIENT_PORT || 4000; //change this to 3000 to run cypress against dev instance
```
Comment (chrisgilmerproj) [inline cypress/support/constants.js]: I need to figure out how to modify this port. Alternatively we make docker-compose run on 4000.

**Comment by chrisgilmerproj** [comment]
What this is missing is the ability to load the e2e test data.

**Comment by chrisgilmerproj** [comment]
Early testing shows that the CI stuff starts running in under 60 seconds (no data is generated for the DB yet!): https://circleci.com/gh/transcom/mymove/147727

I think that's promising for speeding up tests.

File: scripts/run-e2e-test-docker [hunk 2]
```diff
@@ -16,82 +16,123 @@ TSPHOST=tsplocal
...
+  docker-compose down || true
+
+  # Docker Compose Setup
+	aws ecr get-login --no-include-email --region us-west-2 --no-include-email | sh
+	scripts/update-docker-compose
+	docker-compose up -d
+  CONTAINER=mymove_milmove_1
+
+  # Need to wait not just for DB but also migrations to finish
+  DB_TIMEOUT=60 DB_PORT=6432 DB_NAME=dev_db scripts/wait-for-db
```
Comment (chrisgilmerproj) [inline scripts/run-e2e-test-docker]: I constantly forget that CircleCI doesn't let you talk to the running instances via `localhost`. So the only way to talk is via another container you bring up.

File: scripts/run-e2e-test-docker [hunk 3]
```diff
@@ -16,82 +16,123 @@ TSPHOST=tsplocal
...
+    else
+      echo "Waiting 5 seconds for migrations to complete"
+      sleep 5
+    fi
+  done
+
+  # Fill with test data
+  psql postgres://postgres:mysecretpassword@localhost:6432/dev_db?sslmode=disable -c 'TRUNCATE users CASCADE;'
+  make bin/generate-test-data
+  bin/generate-test-data --named-scenario="e2e_basic" --db-env="development" --db-port 6432
```
Comment (chrisgilmerproj) [inline scripts/run-e2e-test-docker]: So while this works locally we'd actually have to do this inside docker to make it work.

**Comment by chrisgilmerproj** [comment]
I'm beginning to think that pushing a new `tools` docker container to AWS makes sense. It would be pretty trivial to do and it would speed things up a lot. We're already building all those tools so why not put them in a useful place?

File: scripts/run-e2e-test-docker [hunk 4]
```diff
@@ -16,82 +16,143 @@ TSPHOST=tsplocal
...
+  # Build a small tools dockerfile
+  make bin_linux/generate-test-data
+  docker build -f Dockerfile.tools --tag tools:latest .
+  docker run \
+    --link="database" \
+    --net "${NETWORK}" \
+    --rm \
+    --entrypoint psql \
+    tools:latest \
+    psql postgres://postgres:mysecretpassword@database:5432/dev_db?sslmode=disable -c 'TRUNCATE users CASCADE;'
```
Comment (chrisgilmerproj) [inline scripts/run-e2e-test-docker]: This is stuck right here. Need to ssh into the container and figure out what's wrong.

Comment (chrisgilmerproj) [reply scripts/run-e2e-test-docker]: Figured it out - you can't specify `psql` both in the entrypoint and the command. doh!

**Comment by chrisgilmerproj** [comment]
At this point I can run e2e tests against docker-compose locally if I modify the `run-e2e-tests-docker` script but I can't seem to get it to work in CircleCI. I think the problem is the migrations are breaking somewhere and I have to figure that part out first.

**Comment by chrisgilmerproj** [comment]
Local migrations file isn't being set properly in CircleCI.

**Comment by chrisgilmerproj** [comment]
Apparently you can't mount folders inside CircleCI with docker: https://circleci.com/docs/2.0/building-docker-images/#mounting-folders

**Comment by chrisgilmerproj** [comment]
It worked! I ran all the tests and 5 failed out of 197! https://circleci.com/gh/transcom/mymove/154215#tests/containers/1

Also, the setup took 11ish minutes instead of 16ish minutes, proving that this is indeed faster!

**Comment by chrisgilmerproj** [comment]
There's some kind of race condition right now with copying in the local migrations folder on CircleCI. Once I solve that this PR is likely ready to go!

**Comment by mr337** [comment]
Kicking off another CI build since it doesn't seem to be related to this PR.

**Comment by chrisgilmerproj** [comment]
> Step 2/10 : COPY bin/rds-combined-ca-bundle.pem /bin/rds-combined-ca-bundle.pem

@garrettqmartin8 Discovered I used `Dockerfile` instead of `Dockerfile.local` for this. You don't need the RDS certs for local development and testing so I'll push out a fix shortly.  If only I'd hit `make clean e2e_test_docker` I would have caught this!

**Comment by mr337** [comment]
`make e2e_test` is working fine and pulls up cypress for further testing

`make e2e_test_docker` runs into the following
```
Creating mymove_milmove_migrate_1 ...
Creating mymove_milmove_migrate_1 ... done
Creating mymove_milmove_1 ...
Creating mymove_milmove_1 ... done
+ docker-compose logs -f milmove_migrate
Attaching to mymove_milmove_migrate_1
milmove_migrate_1  | standard_init_linux.go:211: exec user process caused "no such file or directory"
mymove_milmove_migrate_1 exited with code 1
+ true
++ docker ps --filter=name=mymove_milmove_migrate_1 --filter=status=exited -q
+ MIGRATE_EXIT=b17f6f990bcb
+ [[ -n b17f6f990bcb ]]
+ break
++ docker inspect '--format={{.State.ExitCode}}' mymove_milmove_migrate_1
+ MIGRATE_EXIT_CODE=1
+ [[ 1 != \0 ]]
+ echo

+ echo 'Migration exited with exit code 1'
Migration exited with exit code 1
+ exit 1
Makefile:619: recipe for target 'e2e_test_docker' failed
make: *** [e2e_test_docker] Error 1
```

Granted this is on ubuntu but don't fully understand why `e2e_test` and CircleCI works with no problems. Could be related to my env.

**Comment by mr337** [comment]
Following up on this seeing odd stuff with the binary that I believe is related to the issue I encountered.

```
# calling /bin/milmove
$ docker run  -it -v local_migrations:/migrate/local_migrations --entrypoint /bin/milmove mymove_milmove_migrate
standard_init_linux.go:211: exec user process caused "no such file or directory"
failed to resize tty, using default size

# calling /bin/BLAH as a test
$ docker run  -it -v local_migrations:/migrate/local_migrations --entrypoint /bin/BLAH mymove_milmove_migrate
docker: Error response from daemon: OCI runtime create failed: container_linux.go:345: starting container process caused "exec: \"/bin/BLAH\": stat /bin/BLAH: no such file or directory": unknown.
```

Also checking the container the milmove ELF binary is present. Can head, check size, but cannot run which is really weird.

```
$ docker run  -it -v local_migrations:/migrate/local_migrations --entrypoint /bin/sh mymove_milmove_migrate
/ # cd bin
/bin # ls
ash            dnsdomainname  hostname      *milmove*       pipe_progress  stat
base64         dumpkmap       ionice         mkdir          printenv       stty
bbconfig       echo           iostat         mknod          ps             su
busybox        ed             ipcalc         mktemp         pwd            sync
cat            egrep          kbd_mode       more           reformime      tar
chgrp          false          kill           mount          rev            touch
chmod          fatattr        link           mountpoint     rm             true
chown          fdflush        linux32        mpstat         rmdir          umount
conspy         fgrep          linux64        mv             run-parts      uname
cp             fsync          ln             netstat        sed            usleep
date           getopt         login          nice           setpriv        watch
dd             grep           ls             pidof          setserial      zcat
df             gunzip         lzop           ping           sh
dmesg          gzip           makemime       ping6          sleep

/bin # bin/milmove
/bin/sh: bin/milmove: not found

/bin # ls -la
lrwxrwxrwx    1 root     root            12 Mar  6 20:18 makemime -> /bin/busybox
-rwxrwxr-x    1 root     root      55090527 Aug 23 19:07 milmove
lrwxrwxrwx    1 root     root            12 Mar  6 20:18 mkdir -> /bin/busybox
lrwxrwxrwx    1 root     root            12 Mar  6 20:18 mknod -> /bin/busybox
```

Though it may be related to how the distroless handles execution of binaries.

**Comment by chrisgilmerproj** [comment]
@mr337 - For you to test in Linux you probably need to set `export CI=true` because I use `docker-compose.yml` in linux instead of `docker-compose.local.yml`. The difference is the primary one pulls images from AWS but the other builds locally. Building locally is fraught with edge cases because I'm trying to build linux binaries and shove them into a container. Also the mounting of directories might work differently for you on Linux (not sure why but docker is weird). Try running this on your mac laptop and see what happens.

**Comment by mr337** [comment]
> @mr337 - For you to test in Linux you probably need to set `export CI=true` because I use `docker-compose.yml` in linux instead of `docker-compose.local.yml`. The difference is the primary one pulls images from AWS but the other builds locally. Building locally is fraught with edge cases because I'm trying to build linux binaries and shove them into a container. Also the mounting of directories might work differently for you on Linux (not sure why but docker is weird). Try running this on your mac laptop and see what happens.

Good to know. I'm not trying to be difficult with the linux setup, just trying things out and hoping to find edge cases/assumptions on OSX vs linux, but wow it has been painful.

**Pull request merged**